lib/aiken-design-patterns/multi-utxo-indexer-one-to-many.ak

```diff
@@ -1,93 +1,493 @@
-use aiken/list
-use aiken/transaction.{Input, Output, ScriptContext, Transaction}
-use aiken/transaction/credential.{Address, ScriptCredential}
-use aiken_design_patterns/stake_validator as stake_validator
+use aiken/builtin
+use aiken/collection/list
+use aiken/crypto.{ScriptHash}
+use aiken/fuzz
+use aiken_design_patterns/stake_validator
+use aiken_design_patterns/utils
+use cardano/address.{Address, Credential, Script}
+use cardano/transaction.{
+  Input, NoDatum, Output, Redeemer, ScriptPurpose, Spend, Transaction,
+  placeholder,
+}
 
-pub fn withdraw(
-  input_validator: fn(Input) -> Bool,
-  input_output_validator: fn(Output, Output) -> Bool,
-  collective_output_validator: fn(List<Output>, Int) -> Bool,
-  redeemer: List<(Int, List<Int>)>,
-  ctx: ScriptContext,
+/// Use this function inside your withdrawal script to validate all the inputs
+/// coming from **the script's spend endpoint**. This is an important detail, as
+/// the validator needs to guarantee an exact number of inputs are spent.
+///
+/// If you want to work with an external staking script, consider
+/// using [`withdraw_with_redeemer`](#withdraw_with_redeemer).
+///
+/// The arguments are:
+/// 1. Validation function on each input, and each of its corresponding outputs
+/// 2. Validation function on each input, against all its outputs
+/// 3. Validation function on all the inputs and all the outputs
+/// 4. `Pairs` of indices, mapping each input to multiple outputs
+/// 5. Total number of inputs
+/// 6. Staking credential of the wrapping validator (provided by `withdraw`).
+///    **Note** that in this variant, it can only validate spending the UTxOs
+///    from its own spending endpoint
+/// 7. `Transaction` provided by the validator
+///
+/// For validation functions, corresponding indices of inputs/outputs are also
+/// provided in these functions.
+pub fn withdraw_no_redeemer(
```

Comment:
Can this function be split to aid readability? (Only if it does not make performance worse...)

Reply:
By splitting, do you mean having different variants, each of which perform one of the 3 validation functions?